nntrainer/models/model_loader.cpp

```diff
@@ -478,63 +473,29 @@ int ModelLoader::loadFromIni(std::string ini_file, NeuralNetwork &model,
     NN_INI_RETURN_STATUS();
   }
 
-  ml_logd("parsing ini started");
-  /** Get all the section names */
-  ml_logi("==========================parsing ini...");
-  ml_logi("not-allowed property for the layer throws error");
-  ml_logi("valid property with invalid value throws error as well");
-  for (int idx = 0; idx < num_ini_sec; ++idx) {
-    std::string sec_name = iniparser_getsecname(ini, idx);
-    ml_logd("probing section name: %s", sec_name.c_str());
-
-    if (sec_name.empty()) {
-      ml_loge("Error: Unable to retrieve section names from ini.");
-      status = ML_ERROR_INVALID_PARAMETER;
-      NN_INI_RETURN_STATUS();
-    }
-
-    if (strncasecmp(model_str, sec_name.c_str(), model_len) == 0)
-      continue;
-
-    if (strncasecmp(dataset_str, sec_name.c_str(), dataset_len) == 0)
-      continue;
-
-    if (strncasecmp(optimizer_str, sec_name.c_str(), optimizer_len) == 0)
-      continue;
-
-    /** Parse all the layers defined as sections in order */
-    std::shared_ptr<Layer> layer;
-
-    /**
-     * If this section is a backbone, load backbone section from this
-     * @note The order of backbones in the ini file defines the order on the
-     * backbones in the model graph
-     */
-    const char *backbone_path =
-      iniparser_getstring(ini, (sec_name + ":Backbone").c_str(), unknown);
-
-    const std::string &backbone = resolvePath(backbone_path);
-    if (backbone_path == unknown) {
-      status = loadLayerConfigIni(ini, layer, sec_name);
-    } else if (fileIni(backbone_path)) {
-      status = loadBackboneConfigIni(ini, backbone, model, sec_name);
-      NN_INI_RETURN_STATUS();
-      continue;
-    } else {
-      status = loadBackboneConfigExternal(ini, backbone, layer, sec_name);
-    }
-    NN_INI_RETURN_STATUS();
+  auto path_resolver = [this](const std::string &path) {
+    return resolvePath(path);
+  };
 
-    status = model.addLayer(layer);
-    NN_INI_RETURN_STATUS();
+  ml_logd("parsing graph started");
+  try {
+    std::unique_ptr<GraphInterpreter> ini_interpreter =
+      std::make_unique<nntrainer::IniGraphInterpreter>(app_context,
+                                                       path_resolver);
+    model.model_graph = *ini_interpreter->deserialize(ini_file);
+  } catch (std::exception &e) {
+    ml_logd("failed to load graph, reason: %s ", e.what());
+    status = ML_ERROR_INVALID_PARAMETER;
+    goto end;
   }
-  ml_logd("parsing ini finished");
+  ml_logd("parsing graph finished");
```

Comment:
Well, I think line 491- 496 can be moved to try block, and you may remove `end`.

Reply:
Done :)